### PR TITLE
Initial attempt at KQM substat optimization implementation

### PR DESCRIPTION
**Pull request by Tsym-or-Tysm**
Currently a completely beta/rough implementation, coded up as a proof of concept.

Performance is currently pretty solid among the configs that I've tested. [This sheet](https://docs.google.com/spreadsheets/d/1ThRAQSCQ0o0kzjyG6bFjYIR98yhGCk32GzlV53PwaQw/edit#gid=0) contains a sampling of the configs that I've tested - so far haven't encountered any issue where this implementation completely bombs out.

image

Tried a few external optimization packages to evaluate performance, but in the end things ended up too slow with difficult general parameter tuning so I manually implemented a lot of stuff which can be done better.

Please give comments/suggestions - I would be totally ok with changing large chunks of this if possible since it's so ugly...

### CODE REVIEW

**Comment by HanSeoulOh** [comment]
well i need to use this so might as well review it

**Comment by HanSeoulOh** [comment]
Ok there is a million outputs now, and it's not clear which one is which anymore. We may need clean it up a little but it is very low priority because there's only a handful of us using the CLI.

File: cmd/gcsim/optimize_substats.go [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+	var data strings.Builder
+	var re = regexp.MustCompile(`(?m)^import "(.+)"$`)
+
+	// Regex to identify main stats based on flower. Check that characters all have one that we can recognize
+	var reMainstats = regexp.MustCompile(`(?m)^[A-z]+ add stats.* hp=(4780|3571).*\n`)
+	if len(reMainstats.FindAll(src, -1)) != 4 {
+		sugarLog.Error("Error: Could not identify valid main artifact stat rows for all characters based on flower HP values.")
+		sugarLog.Error("5* flowers must have 4780 HP, and 4* flowers must have 3571 HP.")
+		os.Exit(1)
+	}
```
Comment (HanSeoulOh) [inline cmd/gcsim/optimize_substats.go:110]:
```suggestion
	var reMainstats = regexp.MustCompile(`(?m)^[A-z]+ add stats.* hp=(4780|3571).*\n`)
	var numChars = regexp.MustCompile(`char`)

	if len(reMainstats.FindAll(src, -1)) != len(numChars.FindAll(src, -1)) {
		sugarLog.Error("Error: Could not identify valid main artifact stat rows for all characters based on flower HP values.")
		sugarLog.Error("5* flowers must have 4780 HP, and 4* flowers must have 3571 HP.")
		os.Exit(1)
	}

```

To allow for substat optimization for less than 4 chars

**Review by HanSeoulOh — CHANGES_REQUESTED** [review]

**Comment by Tsym-or-Tysm** [comment]
Note that there will still need to be cleanup done on this at some point - will re-open another issue for it.

**Pull request merged**